### ISSUE TITLE
Support passing `FnOnce` to JS

### ISSUE THREAD

**Issue by Pauan** [rustwasm/wasm-bindgen#1763]
### Motivation

We currently have support for `Closure<FnOnce>`, but not passing `FnOnce` directly to JS.

The [`Promise::new` function](https://docs.rs/js-sys/0.3.27/js_sys/struct.Promise.html#method.new) currently accepts an `FnMut`, but the closure will only be called once, so it should be `FnOnce` instead.

This has practical implications. Right now the wasm-bindgen-futures code looks like this:

```rust
let mut future = Some(future);

Promise::new(&mut |resolve, reject| {
    let future = future.take().unwrap_throw();

    ...
})
```

But if it were `FnOnce` we could simplify that to this:

```rust
Promise::new(move |resolve, reject| {
    ...
})
```

**Comment by fitzgen** [comment]
FWIW, we have https://docs.rs/wasm-bindgen/0.2.50/wasm_bindgen/closure/struct.Closure.html#method.once

**Comment by Pauan** [comment]
@fitzgen Yeah, I mentioned that above, but that's not what this issue is about.

(Though it might be able to reuse some parts of the implementation of `Closure::once`)

**Comment by haze** [comment]
Any updates on this?

**Comment by mwberry** [comment]
From the linked item #3550, I wanted to point out that the deeper the nesting of the closures, the uglier this gets from an ownership perspective. Having `Promise::new` accept an `FnOnce` resolves a ton of ownership complexity.

I'm curious whether users are typically building JS Promises directly via `Promise::new` or if they're using Rust futures and then converting them to Promises via [wasm_bindgen_futures](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/index.html)::[future_to_promise](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/fn.future_to_promise.html#)?
